# RiboseqTools
 Tools for ribosome profiling data analysis and plot.
 
 ### **Prerequisite**
 
 * bowti2 (http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
 
 * sambamba (http://lomereiter.github.io/sambamba/)
 
 * perl modules
 
   * BioPerl
   * Getopt::Long
   * Statistics::Basic
   * Statistics::R
   * Term::ANSIColor
   * Bio::Tools::CodonTable
   
   

**Useage:**

	perl Riboseq_tools.pl [task] [options]

**Task: map meta plot diff**

External tools required: bowtie2 sambamba 

options:

Global parameters:
	
	-dir      output dir, default: RF_output
	-prefix   prefix of output, default: riboseq.
	-threads  threads number, default: 10.
	
** **map** **
This command creates a modified cds or intron-free genome for reads mapping, 
but only cds sequences are used for downstream analyses.

	-read	  reads file,compressed or not,separated by comma, eg: read1.fq,read2,fq [required]   
	-genome   genome file in fasta format [required]
	-gff      gff file [required]
	-type     mappting to cds or intron-free genome, options: cds,genome. Default: cds
	-ext      N bp upstream and downstream sequences to be added to CDS, default: 45	
	-mapq     minimal mapping quality, default: 20
	-mis      maximal mismatch allowed, default: 1
	-seed     seed length for bowtie2, default: 22

** meta **
This command evaluates the patterns of phasing, reads coverage and codon enrichment.
Figures are generated automatically.

	-bam      sorted bam files, separated by comma, eg: bam1,bam2 [required]
	-cds      cds sequence file generated by 'map' command [required]
	-rdRange  read length ranges for codon analysis, format: num-num. Default: 27-30
	-cdRange  codon range to count, cannot > ("ext"/3), format: neg,pos. Default: -15,15
	-minCount minimal count of a gene to use, default: 20
	
** plot **
Plot data generated by meta command, one bam file one graph (.pdf)
R libraries reqiured: ggplot2, reshape, gridExtra

	-bam      bam files used in meta command, separated by comma, eg: bam1,bam2 [required]
	

