# RiboseqTools
 Tools for ribosome profiling data analysis and plot.
 
 ### **Prerequisite**
 
 * bowti2 (http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
 
 * sambamba (http://lomereiter.github.io/sambamba/)
 
 * perl modules
 
   * BioPerl
   * Getopt::Long
   * Statistics::Basic
   * Statistics::R
   * Term::ANSIColor
   * Bio::Tools::CodonTable

 * Ribodiff (https://github.com/ratschlab/RiboDiff) or Riborex (https://github.com/smithlabcode/riborex)
 * DESeq2
 * R libraries
   * ggplot2
   * reshape
   * gridExtra
 
 ### **Useage:**
     perl Riboseq_tools.pl  [task]  [options]

### Task: 
 * **map**
 * **meta**
 * **plot**
 * **diff**

#### **options:**

**Global parameters:**
	
	-dir      output dir, default: RF_output
	-prefix   prefix of output, default: riboseq.
	-threads  threads number, default: 10.
	
** **map** **

This command creates a modified cds or intron-free genome for reads mapping, 
but only cds sequences are used for downstream analyses.

	-read	  reads file,compressed or not,separated by comma, eg: read1.fq,read2,fq [required]   
	-genome   genome file in fasta format [required]
	-gff      gff file [required]
	-type     mappting to cds or intron-free genome, options: cds,genome. Default: cds
	-ext      N bp upstream and downstream sequences to be added to CDS, default: 45	
	-mapq     minimal mapping quality, default: 20
	-mis      maximal mismatch allowed, default: 1
	-seed     seed length for bowtie2, default: 22

** **meta** **

This command evaluates the patterns of data, eg: phasing, reads coverage and codon enrichment.
Figures are generated automatically.

	-bam      sorted bam files, separated by comma, eg: bam1,bam2 [required]
	-cds      cds sequence file generated by 'map' command [required]
	-rdRange  read length ranges for codon analysis, format: num-num. Default: 27-30
	-cdRange  codon range to count, cannot > ("ext"/3), format: neg,pos. Default: -15,15
	-minCount minimal count of a gene to use, default: 20
	
** **plot** **

Plot data generated by **meta** command, one bam file one graph (.pdf)

	-bam      bam files used in meta command, separated by comma, eg: bam1,bam2 [required]
	
** **diff** **

This comman calculates differential expression/translation using DESeq2 (for mRNA/RFP) or Ribodiff (for TE)
TE can also be calculated using **Riborex**, which is faster but cannot be applied when the number
of samples are different for mRNA and RFP.

        -cds      cds sequence file generated by 'map' command [required]
        -rfLen    reads within this range to count for Riboseq, default: 27-30
        -rsLen    reads within this range to count for mRNAseq, default: 27-30
        -strand   mappng strand to count: +: 0; -: 1; both: 2. Default: 2
        -engine   tool to be used for differential TE analysis, ribodiff or riborex. Default: ribodiff
        -bamRFCnt bam files of riboseq used as control, separated by comma. [required]
        -bamRFTrt bam files of riboseq used as treatmeat. [required]
        -bamRSCnt bam files of mRNAseq used as control. [required]
        -bamRSTrt bam files of mRNAseq used as treatmeat. [required]
        -rdmin    minimal sum of normalized reads for the test. default: 10

options for **ribodiff**

        -src      location for ribodiff script (TE.py), defualt: utils/TE.py
        -disp     dispersion for riboseq or mRNAseq. 1: different dispersion; 0: same dispersion. default
        -padj     method for multiple test correction, BH, Bonferroni. Default: BH
        -plot     make plots to show the data and results. On: 1; Off: 0. Defulat: 0
        -minP     FDR cutoff for plot. Default: 0.1

options for **riborex**

        -tool     tools for estimation, options: DESeq2, edgeR, edgeRD, Voom. Default: DESeq2.
	
